docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,456 @@
+    - If the `checked` parameter is set to `true`, the [xor()](#xor) gadget is reused with a second argument to be an all-ones bitmask of the same length. This approach is less efficient because it needs as many rows as an XOR requires for a single negation, but comes with the advantage of making sure the input is of a certain length.
+
+For implementation details, see [NOT](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#not) in the Mina book.
+
+Example NOT-ing 4 bits with the unchecked version:
+
+```ts
+let a = Field(0b0101);
+let b = Gadgets.not(a,4,false);
+
+b.assertEquals(0b1010);
+```
+
+Example NOT-ing 4 bits with the checked version using `xor()`:
+
+```ts
+let a = Field(0b0101);
+let b = Gadgets.not(a,4,true);
+
+b.assertEquals(0b1010);
+```
+
+## xor()
+
+```ts
+xor(a: Field, b: Field, length: number) => Field
+```
+
+The `xor()` gadget is equivalent to the [Bitwise XOR (^)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR) operator in JavaScript. It is a provable method to support bitwise XOR operations for native Field elements. 
+
+The XOR operation:
+
+- Compares two bits and returns 1 if two bits differ. Returns 0 if the two bits are equal.
+- Recursively builds a chain of XOR gates. 
+- Verifies that each XOR gate can verify at most 16 bits.
+- Adds another XOR gate to the chain if your input elements exceed 16 bits.
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- Constrains input values to fit into `paddedLength` bits. 
+- Requires that both Field elements fit into `2^paddedLength`.
+    For example, `length = 2` (`paddedLength = 16`) fails for any input that is larger than `2**16`.
+- Guarantees the output has at most `paddedLength` bits.
+
+The `length` parameter defines how many bits to compare. When choosing the value of the `length` parameter, be sure to carefully weigh the trade-off between an increased number of constraints and security:
+- A larger `length` parameter adds more constraints.
+- A smaller `length` allows the verifier to infer the length of the original input data. For example, it will be smaller than 16 bits if only one XOR gate is used.
+
+For details about the implementation, see [XOR](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#xor-1) in the Mina book.
+
+Example:
+
+```ts
+let a = Field(0b0101);
+let b = Field(0b0011);
+
+let c = Gadgets.xor(a, b, 4); // xor-ing 4 bits
+c.assertEquals(0b0110);
+```
+
+
+## addMod32()
+
+```ts
+addMod32(a: Field, b: Field) => Field
+```
+
+The `addMod32()` operation adds two Field elements in the range `[0, 2^64]` and returns the result modulo `2^32`. It is a helper method to interact with 3-limb vectors of Fields.
+
+The operation:
+
+- Is addition modulo `2^32`.
```

Comment:
Do we need to describe what this is?

Reply:
This is actually (perhaps surprisingly) specific. I suspect that people using this will understand it. afaik it's an efficient way to do addition that is overflow safe. All addition in SHA256 works this way. Will defer to eng. review on whether there's an easy way to explain it here cos that would be nice but I don't think we should go into depth on it here.